src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs

```diff
@@ -155,60 +215,76 @@ internal void InitializeForSourceGen(JsonSerializerOptions options, JsonProperty
             Type declaringType = propertyInfo.DeclaringType;
 
-            JsonConverter? converter = propertyInfo.Converter;
-            if (converter == null)
+            JsonConverter<T>? typedCustomConverter = propertyInfo.Converter;
+            CustomConverter = typedCustomConverter;
+
+            JsonConverter<T>? typedNonCustomConverter = propertyTypeInfo.Converter as JsonConverter<T>;
+            NonCustomConverter = typedNonCustomConverter;
+            JsonConverter<T>? typedEffectiveConverter = typedCustomConverter ?? typedNonCustomConverter;
+            if (typedEffectiveConverter == null)
             {
-                converter = propertyTypeInfo.PropertyInfoForTypeInfo.ConverterBase as JsonConverter<T>;
-                if (converter == null)
-                {
-                    throw new InvalidOperationException(SR.Format(SR.ConverterForPropertyMustBeValid, declaringType, ClrName, typeof(T)));
-                }
+                throw new InvalidOperationException(SR.Format(SR.ConverterForPropertyMustBeValid, declaringType, ClrName, typeof(T)));
             }
 
-            ConverterBase = converter;
-
             if (propertyInfo.IgnoreCondition == JsonIgnoreCondition.Always)
             {
                 IsIgnored = true;
-                Debug.Assert(!ShouldSerialize);
-                Debug.Assert(!ShouldDeserialize);
+                Debug.Assert(!CanSerialize);
+                Debug.Assert(!CanDeserialize);
             }
             else
             {
                 Get = propertyInfo.Getter!;
                 Set = propertyInfo.Setter;
-                HasGetter = Get != null;
-                HasSetter = Set != null;
                 JsonTypeInfo = propertyTypeInfo;
                 DeclaringType = declaringType;
                 IgnoreCondition = propertyInfo.IgnoreCondition;
                 MemberType = propertyInfo.IsProperty ? MemberTypes.Property : MemberTypes.Field;
-                ConverterStrategy = Converter!.ConverterStrategy;
+                ConverterStrategy = typedEffectiveConverter.ConverterStrategy;
                 NumberHandling = propertyInfo.NumberHandling;
             }
         }
 
-        internal override void Configure()
+        internal override void Configure(JsonTypeInfo typeInfo)
         {
-            base.Configure();
+            base.Configure(typeInfo);
 
             if (!IsForTypeInfo && !IsIgnored)
             {
-                _converterIsExternalAndPolymorphic = !ConverterBase.IsInternalConverter && PropertyType != ConverterBase.TypeToConvert;
+                _converterIsExternalAndPolymorphic = !EffectiveConverter.IsInternalConverter && PropertyType != EffectiveConverter.TypeToConvert;
                 _propertyTypeEqualsTypeToConvert = typeof(T) == PropertyType;
             }
         }
 
-        internal override JsonConverter ConverterBase
+        internal override void SetEffectiveConverter()
```

Comment:
Set implies a setter, so maybe use something like
```suggestion
        internal override void DetermineEffectiveConverter()
```